File: CodeWars/4kyu/Snail/solution.py
def snail(array):
    
    if not array:
        return array
    #make a list to return at the end
    out = list()
    n = len(array)
    
    #add first row of array to out
    out += array.pop(0)
    
    #add final column of snailed array to out
    for i in range(1,n-1):
        out.append(array[i-1].pop())
    if array:
        out += array.pop()[::-1]
        
    
    for i in range(n-2,0,-1):
        out.append(array[i-1].pop(0))
        
    #recursively repeat with next "snail" of array
    out += snail(array)
    return out

File: CodeWars/4kyu/Snail/solution.js
snail = function (arr) {
  let out = [];

  //base case for recursion
  if (!arr.length) return out;

  //add top of square to out
  let next = arr.shift();
  if (next) {
    out = out.concat(next);
  }

  //add right hand column of square to out
  for (let i = 0; i < arr.length; i++) {
    out.push(arr[i].pop());
  }

  //add bottom of square to out
  next = arr.pop();
  if (next) {
    out = out.concat(next.reverse());
  }

  //add left hand column of square to out
  for (let i = arr.length - 1; i >= 0; i--) {
    out.push(arr[i].shift());
  }

  //recursively add next square (if any) to out
  return out.concat(snail(arr));
};
